### ISSUE TITLE
`hsup` not behaving right for polymorphic `LNL` grades

### ISSUE THREAD

**Issue by dorchard** [granule-project/granule#168]
When we have a polymorphic LNL grade we should not be able to push. Negative test:
https://github.com/granule-project/granule/blob/dev-minor/frontend/tests/cases/negative/hsup/pushPolyLNL.gr
But this passes the type checker.

**Comment by jackohughes** [comment]
Seems to be happening because a compiled variable LNL grade will have the symbolic grade `(SLNL <SInteger 0 \/ SInteger 1 \/ SInteger 2>)` i.e. a symbolic grade representing the possible SLNL values (see line 228 in Constraints.hs). This means that when we do the check in `symGradeHsup` in SymbolicGrades.hs that grade is equal to the symbolic representation of 1, this passes because what we're really doing `(0 \/ 1 \/ 2) == 1`. 

We can fix this by requiring that the symbolic grades we pass to `symGradeHsup` have a concrete representation. This will be the case if their grade in the type is concrete and not a variable. If it's the case that these grades aren't concrete (i.e. they are variables) it means they're polymorphic within LNL so we don't want to allow hsup in this case (See commit https://github.com/granule-project/granule/compare/0865b0594327...094eaebae8dd)

**Comment by dorchard** [comment]
We've uncovered that there are other issues at play here. This should be allowed but isn't now:
```
good : forall {l : LNL} . {l == One} => ((Int, Int) [l] -> (Int [l], Int [l]))
good [(x, y)] = ([x], [y])
```

**Comment by dorchard** [comment]
This is fixed by 045f4c64
